prime-router/src/main/kotlin/tokens/Scope.kt

```diff
@@ -72,19 +114,93 @@ class Scope {
             val splits = scope.split(".")
             val orgName = splits[0]
             val senderOrReceiver = splits[1]
-            val detailedScope = DetailedScope.valueOf(splits[2])
+            val detailedScope: DetailedScope = DetailedScope.toDetailedScope(splits[2])!! // checked in isValidScope
             return Triple(orgName, senderOrReceiver, detailedScope)
         }
 
         /**
+         * Finds the set of scopes that the user is authorized to have.
+         * Returns the matches between the set of scopes the user claims ([userScopeList])
+         * and the set of scopes acceptable to access a resource ([requiredScopeList])
+         *
+         * @return the (possibly empty) intersection of the two sets.  Empty Set == not authorized for anything.
+         */
+        fun authorizedScopes(userScopes: Set<String>, requiredScopes: Set<String>): Set<String> {
+            return userScopes.filter { it.isNotBlank() }.intersect(requiredScopes.filter { it.isNotBlank() }.toSet())
+        }
+
+        /**
+         * @return true if there is at least one match between the set of scopes the user claims ([userScopeList])
+         * and the set of scopes acceptable to access a resource ([requiredScopeList])
+         */
+        fun authorized(userScopeList: Set<String>, requiredScopeList: Set<String>): Boolean {
+            val intersection = authorizedScopes(userScopeList, requiredScopeList)
+            return if (intersection.isEmpty()) {
+                logger.warn(
+                    "User not authorized. User has scope: (${userScopeList.joinToString(",")});" +
+                        " required scope is: (${requiredScopeList.joinToString(",")})"
+                )
+                false
+            } else {
+                logger.info("User authorized for scopes (${intersection.joinToString(",")})")
+                true
+            }
+        }
+
+        /**
```

Comment:
This function is used by the TokenFunction code to lookup scopes in the Settings, which are associated with public keys.